hydro_lang/src/viz/json.rs

```diff
@@ -0,0 +1,1119 @@
+            .into_iter()
+            .map(|def| {
+                serde_json::json!({
+                    "type": def["id"],
+                    "label": def["label"]
+                })
+            })
+            .collect()
+    }
+
+    /// Get edge style configuration with semantic→style mappings.
+    fn get_edge_style_config() -> serde_json::Value {
+        serde_json::json!({
+            "semanticPriorities": [
+                ["Unbounded", "Bounded"],
+                ["NoOrder", "TotalOrder"],
+                ["Keyed", "NotKeyed"],
+                ["Network", "Local"]
+            ],
+            "semanticMappings": {
+                // Network communication group - controls line pattern AND animation
+                "NetworkGroup": {
+                    "Local": {
+                        "line-pattern": "solid",
+                        "animation": "static"
+                    },
+                    "Network": {
+                        "line-pattern": "dashed",
+                        "animation": "animated"
+                    }
+                },
+
+                // Ordering group - controls waviness
+                "OrderingGroup": {
+                    "TotalOrder": {
+                        "waviness": "straight"
+                    },
+                    "NoOrder": {
+                        "waviness": "wavy"
+                    }
+                },
+
+                // Boundedness group - controls halo
+                "BoundednessGroup": {
+                    "Bounded": {
+                        "halo": "none"
+                    },
+                    "Unbounded": {
+                        "halo": "light-blue"
+                    }
+                },
+
+                // Keyedness group - controls vertical hash marks on the line
+                "KeyednessGroup": {
+                    "NotKeyed": {
+                        "line-style": "single"
+                    },
+                    "Keyed": {
+                        "line-style": "hash-marks"
+                    }
+                },
+
+                // Collection type group - controls color
+                "CollectionGroup": {
+                    "Stream": {
+                        "color": "#2563eb",
+                        "arrowhead": "triangle-filled"
+                    },
+                    "Singleton": {
+                        "color": "#000000",
+                        "arrowhead": "circle-filled"
+                    },
+                    "Optional": {
+                        "color": "#6b7280",
+                        "arrowhead": "diamond-open"
+                    }
+                },
+            },
+            "note": "Edge styles are now computed per-edge using the unified edge style system. This config is provided for reference and compatibility."
+        })
+    }
+
+    /// Optimize backtrace data for size efficiency
+    /// 1. Remove redundant/non-essential frames
```

Comment:
Only doing the first right?

Also, actually, we may want to keep around memory addresses for situations where function `a` calls function `b` twice on the same line (we would want each instance of `b` to be grouped separately). Okay to do in a followup PR.